### ISSUE TITLE
Non-ascii characters (unicode, non-latin or accented characters, emojis etc.) broken after Perl upgrade

### ISSUE THREAD

**Issue by trygveaa** [weechat/weechat#1996]
## Bug summary

After upgrading to Perl 5.38 non-ascii characters are not visible anymore. This may also lead to various rendering artifacts, such as characters not being cleared when you switch buffer, alignment issues and generally characters appearing out of place.

## Steps to reproduce

1. Have Perl 5.38 installed (and WeeChat compiled against it)
2. Load a Perl script or run `/perl eval`
3. Observe issues

## Additional information

The issue occurs because Perl sets the thread locale to `C`. I've opened an issue about it here: https://github.com/Perl/perl5/issues/21366

It seems like something that should be fixed in Perl, but we'll have to wait and hear what they say.

---

- WeeChat version: any
- OS, distribution and version: any (at least any linux)
- Terminal: any
- Terminal multiplexer (screen/tmux/…/none): any

**Comment by trygveaa** [comment]
Here's a workaround to temporarily fix the locale:

```
/perl eval use POSIX (); POSIX::setlocale(&POSIX::LC_ALL, "");
```

After running that, non-acii characters will work again until the next time you load a Perl script.

**Comment by mtasaka** [comment]
Maybe the following can be a workaround?? (I just checked testsuite can pass with the following change):

```
diff --git a/src/plugins/perl/weechat-perl.c b/src/plugins/perl/weechat-perl.c
index 1759a4810..11f652ecc 100644
--- a/src/plugins/perl/weechat-perl.c
+++ b/src/plugins/perl/weechat-perl.c
@@ -567,6 +567,7 @@ weechat_perl_load (const char *filename, const char *code)
     temp_script.interpreter = (PerlInterpreter *) perl_current_interpreter;
     perl_parse (perl_current_interpreter, weechat_perl_api_init,
                 perl_args_count, perl_args, NULL);
+    Perl_setlocale(LC_CTYPE, "");
     length = strlen (perl_weechat_code) + strlen (str_warning) +
         strlen (str_error) - 2 + 4 + strlen ((code) ? code : filename) + 4 + 1;
     perl_code = malloc (length);
@@ -1290,6 +1291,7 @@ weechat_plugin_init (struct t_weechat_plugin *plugin, int argc, char *argv[])
     perl_construct (perl_main);
     perl_parse (perl_main, weechat_perl_api_init, perl_args_count,
                 perl_args, NULL);
+    Perl_setlocale(LC_CTYPE, "");
 #endif /* MULTIPLICITY */
 
     perl_data.config_file = &perl_config_file;
```

**Comment by flashcode** [comment]
Another patch proposed by @trygveaa: https://github.com/trygveaa/weechat/commit/6c86e160713e60fd4d6998cdfb14fdfdc770d8bb

**Comment by trygveaa** [comment]
I think @mtasaka's patch might be better than mine. Maybe we should just apply that since we don't seem to get any responses in the Perl issue?

**Comment by flashcode** [comment]
@mtasaka: thanks for your patch, merged in branches master and 4.0.

I just added a condition on the Perl version (>= 5.38.0), because I still build WeeChat on old distros like Ubuntu Bionic, with Perl 5.26.1 that doesn't provide the function Perl_setlocale.
